src/StreamUtils.bsv

```diff
@@ -0,0 +1,380 @@
+
+    Reg#(Bool) hasRemainReg <- mkReg(False);
+    Reg#(Bool) hasLastRemainReg <- mkReg(False);
+    Reg#(Bool) isStreamAEnd <- mkReg(False);
+
+    Reg#(DataStream) remainStreamReg <- mkRegU;
+    
+
+    rule prepareStreamA;
+        let streamA = inputFifoA.first;
+        inputFifoA.deq;
+        DataBytePtr bytePtr = convertByteEn2BytePtr(streamA.byteEn);
+        prepareFifoA.enq(StreamWithPtr {
+            stream: streamA,
+            bytePtr: bytePtr
+        });
+    endrule
+
+    rule prepareStreamB;
+        let streamB = inputFifoB.first;
+        inputFifoB.deq;
+        DataBytePtr bytePtr = convertByteEn2BytePtr(streamB.byteEn);
+        prepareFifoB.enq(StreamWithPtr {
+            stream: streamB,
+            bytePtr: bytePtr
+        });
+    endrule
+
+    rule concatStream;
+        // Only the remain data
+        if (hasRemainReg && hasLastRemainReg) begin
+            outputFifo.enq(remainStreamReg);
+            hasRemainReg <= False;
+            isStreamAEnd <= False;
+        end
+        // StreamB or streamB + the remain data
+        else if (prepareFifoB.notEmpty && isStreamAEnd) begin
+            let streamB = prepareFifoB.first.stream;
+            let bytePtrB = prepareFifoB.first.bytePtr;
+            prepareFifoB.deq;
+            streamB.isFirst = False;
+            if (hasRemainReg) begin
+                let {concatStream, remainStream, remainBytePtr} = getConcatStream(remainStreamReg, streamB, remainBytePtrReg, bytePtrB);
+                hasRemainReg     <= !isByteEnZero(remainStream.byteEn);
+                hasLastRemainReg <= streamB.isLast;
+                remainStreamReg  <= remainStream;
+                remainBytePtrReg <= remainBytePtr;
+                outputFifo.enq(concatStream);
+            end
+            else begin
+                outputFifo.enq(streamB);
+            end
+            isStreamAEnd <= !streamB.isLast;
+        end
+        // StreamA or StreamA + first StreamB
+        else if (prepareFifoA.notEmpty) begin
+            let streamA = prepareFifoA.first.stream;
+            let bytePtrA = prepareFifoA.first.bytePtr;
+            // Only StreamA frame
+            if (!streamA.isLast) begin
+                outputFifo.enq(streamA);
+                prepareFifoA.deq;
+                isStreamAEnd <= False;
+            end 
+            // the last StreamA + the first StreamB
+            else if (streamA.isLast && prepareFifoB.notEmpty) begin
+                let streamB = prepareFifoB.first.stream;
+                let bytePtrB = prepareFifoB.first.bytePtr;
+                let {concatStream, remainStream, remainBytePtr} = getConcatStream(streamA, streamB, bytePtrA, bytePtrB);
+                hasRemainReg     <= !isByteEnZero(remainStream.byteEn);
+                hasLastRemainReg <= streamB.isLast;
+                remainStreamReg  <= remainStream;
+                remainBytePtrReg <= remainBytePtr;
+                isStreamAEnd     <= !streamB.isLast;
```

Comment:
`isStreamAEnd` is misleading. `isStreamAEnd` looks like `streamA.isLast`,  but infact it's not.

Reply:
`isStreamAEnd` means `streamA.isLast`, but it need to be reset when the whole concat ends, i.e.  `streamB.isLast` asserts. The codes here are misleading and I have modified for better understanding.